### PR TITLE
ci: use vocone

### LINKED ISSUE #6: bug: use local vocone in CI, instead of remote API
if i understand the issue correctly:
the CI is currently hitting a realworld API (`secrets.API_URL`)
https://github.com/vocdoni/vocdoni-sdk/blob/088ced66cc98ffcf98bdc0424a6495318662918f/.github/workflows/main.yml#L34-L37
while it should spin up a ephemeral vochain using vocone on docker, and hit that instead.

### PR DESCRIPTION
**Pull request by altergui**
fix #6

### CODE REVIEW

**Comment by altergui** [comment]
well, here's my stab at this problem. I managed to run voconed cleanly inside a docker and set the API_URL.

the test does fail with status code 405 Method not allowed, but from my understanding this is due to the fact that voconed doesn't implement the APIv2.

so i'm not sure how to proceed. TBH i struggled with voconed since for example `--treasurer` arg was only half-implemented, can be passed as cmdline arg but not as env var (as the rest of options), so it did give me the impression voconed it's a bit unused? and thus unmaintained?

if that's the case, then the sensible option (overkill?) would be to spin up everything with dvotenode, i.e. a whole testsuite network (miners, gateways with apiv2)

**Comment by altergui** [comment]
so the sensible path forward is to port the apiv2 support to voconed, since it's not abandoned nor deprecated.

**Comment by altergui** [comment]
rebased on current main,
currently making some improvements

**Comment by altergui** [comment]
so i'd say this is ready for review... but probably not for merging since it needs work on election.test.ts and accounts.test.ts to really hit the local voconed instance and not the remote public API

https://github.com/vocdoni/vocdoni-sdk/actions/runs/3657707246/jobs/6181600196
```
PASS test/integration/account.test.ts (68.878 s)
PASS test/integration/other.test.ts
FAIL test/integration/election.test.ts (560.334 s)
```
account.test.ts passes just fine (but i believe hitting the remote API, not voconed, even though API_URL is correctly set to hit voconed, i guess there's a bug in env var precedence?)
and election.test.ts times out, for whatever reason.

voconed runs just fine but gets no processes or votes.
```
voconed_1  | 2022-12-09T13:29:20Z	INFO	vocone/vocone.go:477	[vochain info] height:279 mempool:0 processes:0 votes:0 vxm:0 voteCache:0 blockTime:{1m:2.00}
```
i don't know who would do that (AFAIU @marcvelmer wouldn't, focused on other tasks) and I certainly wouldn't either with zero proficiency in TypeScript

**Comment by altergui** [comment]
i'm leaving this in a mergeable state (disabled `election.test.ts`) that doesn't break anything, so you can merge it if you want as-is and let someone else work further on this feature outside of this PR

File: .github/workflows/main.yml [hunk 1]
```diff
@@ -30,3 +30,58 @@ jobs:
...
+      - name: Start voconed (ephemeral vochain for integration test)
+        run: |
+          docker-compose -f test/integration/docker-compose.yml up -d
+          echo -n "voconed_hostport=" >> $GITHUB_ENV
+          docker-compose -f test/integration/docker-compose.yml port voconed 9095 >> $GITHUB_ENV
+
+      - name: Integration tests
+        run: yarn test:integration --ci --coverage --maxWorkers=2
+        env:
+          API_URL: http://${{ env.voconed_hostport }}/dvote
```
Comment (p4u) [inline .github/workflows/main.yml:76]: This looks incorrect. The default API route is /api

File: test/integration/voconed.env [hunk 2]
```diff
@@ -0,0 +1,5 @@
+VOCONED_PORT=9095
+VOCONED_URLPATH=/dvote
```
Comment (p4u) [inline test/integration/voconed.env:2]: Ah, I see you'r changing the route here. I'd not use dvote anymore, but /api or /v2 (the last one is the default for vocodoni-node)

**Comment by p4u** [comment]
> @p4u suggested:
> 
> 1. either skip account creation (and faucet) and do the tests assuming the account 0x3e951345Fd7f3fE62606356C31a1a7CD948ea101 exists and has 1000 tokens balance (which happens during voconed init, given the flags passed in docker-compose.yml)
> 2. alternatively, make that account a faucet and pre-generate a FaucetPkg to be hardcoded in the tests
> 3. last resource: set all TxCosts to 0 so no tokens are needed, but that would severely decrease test coverage, not a good idea.
> 
> right now this PR allows option 1. for option 2 it would need a slight adapting.

the second options is easy to do. There are helper functions in vocdoni-node that help create the faucet package (we only need to do it once).

File: test/integration/voconed.env (hunk 2 above)
Comment (altergui) [reply test/integration/voconed.env]: done, cf697616880e0996492f051898aa18537e7de235 uses `/v2` now

**Comment by marcvelmer** [comment]
@altergui I tried to run it but it fails creating the election. In the logs I saw this response, which has the block times to 0, and they are needed in order to calculate the end block end for the process.

`{"chainId":"vocone","blockTime":[0,0,0,0,0],"height":5,"blockTimestamp":1671486197}`

**Comment by altergui** [comment]
digging further into this, (election.test.ts failing) i have the impression it's not hitting the local vocone instance
https://github.com/vocdoni/vocdoni-sdk/actions/runs/3735721080/jobs/6339277794

there are some logged POST to `/elections` endpoint, for example
https://github.com/vocdoni/vocdoni-sdk/actions/runs/3735721080/jobs/6339277794#step:10:241
but all of them resulted in 200 OK, no trace of error 400 from vocone. that leads me to think the test is hitting the remote API, where the process does in fact not exist. could this be somehow the case?

**Comment by altergui** [comment]
indeed, hunch correct. this client lacks api_url and hits the default (remote) api

https://github.com/vocdoni/vocdoni-sdk/blob/5a2ec51f5b9cee3a21aab3a6d21bdd3acc7a2618/test/integration/election.test.ts.disabled#L127-L130

_gui typescript intensifies_

**Comment by altergui** [comment]
yay! something happened. on one hand, it seems account.test.ts was still hitting the public api and now, hitting API_URL, runs much faster (21 s vs ~75 s)

election.test.ts still fails but with a different error, so it seems my fix in  d742c3aa49630ad532c50a73d71364941cb1c54c was correct

```
PASS test/integration/account.test.ts (21.886 s)
PASS test/integration/other.test.ts
FAIL test/integration/election.test.ts (440.725 s)
  ● Election integration tests › should create an election with 100 participants and each of them should vote correctly

    Request error: socket hang up

      44 |       .catch((error) => {
      45 |         if (axios.isAxiosError(error)) {
    > 46 |           throw new Error('Request error: ' + error.message);
         |                 ^
      47 |         }
      48 |         throw error;
      49 |       });

      at src/api/vote.ts:46:17
          at runMicrotasks (<anonymous>)
      at VocdoniSDKClient.submitVote (src/client.ts:372:22)
          at async Promise.all (index 44)
      at Object.<anonymous> (test/integration/election.test.ts:117:5)

  ● Election integration tests › should create an election with 100 weighted participants and each of them should vote correctly

    Request error: socket hang up

      44 |       .catch((error) => {
      45 |         if (axios.isAxiosError(error)) {
    > 46 |           throw new Error('Request error: ' + error.message);
         |                 ^
      47 |         }
      48 |         throw error;
      49 |       });

      at src/api/vote.ts:46:17
          at runMicrotasks (<anonymous>)
      at VocdoniSDKClient.submitVote (src/client.ts:372:22)
          at async Promise.all (index 0)
      at Object.<anonymous> (test/integration/election.test.ts:162:5)

  ● Election integration tests › should create an encrypted election with 100 participants and each of them should vote correctly

    Request error: socket hang up

      44 |       .catch((error) => {
      45 |         if (axios.isAxiosError(error)) {
    > 46 |           throw new Error('Request error: ' + error.message);
         |                 ^
      47 |         }
      48 |         throw error;
      49 |       });

      at src/api/vote.ts:46:17
          at runMicrotasks (<anonymous>)
      at VocdoniSDKClient.submitVote (src/client.ts:372:22)
          at async Promise.all (index 94)
      at Object.<anonymous> (test/integration/election.test.ts:206:5)
```

**Review by marcvelmer — APPROVED** [review]
Already merged to `main`

**Comment by marcvelmer** [comment]
Already merged at 0c173743749a78cc34b4adae86cc7f4efc4298b9

**Pull request closed without merging**